File: solutions/csharp/12.cs
using System;

public class Twelve
{
	static void Main()
	{
		long a = 1;
		long b = 1;
		long c = 0;

		while (c <= 500)
		{
			c = factors(a);

			if (c > 500)
			{
				Console.Write("\n{0}\n", a);
				break;
			}
			else
			{
				a += b + 1;
				b++;
			}
		}
	}

	static long factors(long a)
	{
		long b = 1;

		for (int i = 1; i <= Math.Sqrt(a); i++)
			if (a % i == 0)
				b++;

		return 2 * b;
	}
}

File: javascript/12.js
function divisorCount(number) {
    let end = Math.ceil(Math.sqrt(number));
    let divisorCount = 0;

    for (let i = 1; i < = end; i++) {
        if (!(number % i)) {
            divisorCount += 2;
        }
    }

    return divisorCount;
}

function* triangleGenerator(number) {
    let current = 0;

    while (true) {
        current += 1;
        yield (current + 1) * current >> 1;
    }
}

function solver(goal, generator) {
    const triangle = generator.next().value;

    if (divisorCount(triangle) > goal) {
        return triangle;
    } else {
        return solver(goal, generator);
    }
}

console.log(solver(500, triangleGenerator()));